sdk/identity/azure-identity/src/chained_token_credential.cpp

```diff
@@ -0,0 +1,35 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include "azure/identity/chained_token_credential.hpp"
+
+#include <utility>
+
+using namespace Azure::Identity;
+
+ChainedTokenCredential::ChainedTokenCredential(ChainedTokenCredential::Sources sources)
+    : m_sources(std::move(sources))
+{
+}
+
+ChainedTokenCredential::~ChainedTokenCredential() = default;
+
+Azure::Core::Credentials::AccessToken ChainedTokenCredential::GetToken(
+    Azure::Core::Credentials::TokenRequestContext const& tokenRequestContext,
+    Azure::Core::Context const& context) const
+{
+  for (const auto& source : m_sources)
+  {
+    try
+    {
+      return source->GetToken(tokenRequestContext, context);
+    }
+    catch (const Azure::Core::Credentials::AuthenticationException&)
```

Comment:
It seems like we want to throw AuthException if none of the sources work. Should we try n-1 sources, with the try/catch, and the try the last one outside the catch block to propogate it's exception directly?

I am not sure what the best solution is here, and maybe I am mistaken, but I thought a try/catch and re-throw in C++ is a bit of a code smell, because you lose the original stack and actual source of exception. @CaseyCarter ?

Reply:
My gut response to this design is to think that we should have a `TryGetToken` operation that returns a token on success and failure info otherwise, with a means to aggregate failure info from multiple sources to provide to the caller (possibly via exception) in the event they all fail. It seems like we're expecting failures to occur frequently here, as opposed to that being an exceptional event.

Reply:
We can't have `TryGetToken()`, because it is an override function that we are implementing here.
I don't think we need to go that far, especially given that other SDKs don't have it.

@ahsonkhan, if you rethrow with `catch(ex){ throw; }` and not `catch(ex){ throw ex; }`, you don't lose the call stack.

I don't think we should throw only the last one. But we could.

We also could derive from AuthException and create some sort of exception which holds a full list of exceptions, and it won't be a breaking change. None of it would be a breaking change, even if we decide to throw the last one, so let's collect some feedback?

But I don't think it is needed. As a developer, the actions you take to recover are very likely the same, be the auth error from credential `#1`, or `#2`.

Plus I just added logging, so you'll be able to get errors from the log.